Implement createConnectionConsumer for Jboss EE . - Required for Consuming messages for Jboss Container

Implement createConnectionConsumer for Jboss EE . This is the method called by the JBoss container.

This method is currently not implemented https://github.com/datastax/pulsar-jms/blob/96606d0f22a1af8fdde0c5eff0f5dde086d9862c/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnection.java#L685

Hello @eolivelli,  
It seems to move a little ahead and called the ConnectionConsumer but gave a null pointer. It required me to put webserviceURL also - which is not required while publishing...but required for consuming .
Everything worked with PR except one compilation error.
I just changed word "apply" to "build" in PulsarConnection line no. 966   MessageConsumer consumer = consumerBuilder.build(session);

Error Stack 
javax.jms.JMSException: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
        at com.datastax.oss.pulsar.jms.Utils.handleException(Utils.java:63)
        at com.datastax.oss.pulsar.jms.PulsarConnectionFactory.ensureQueueSubscription(PulsarConnectionFactory.java:963)
        at com.datastax.oss.pulsar.jms.PulsarMessageConsumer.subscribe(PulsarMessageConsumer.java:116)
        at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:783)
        at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:734)
        at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:73)
        at com.datastax.oss.pulsar.jms.PulsarConnection.lambda$createConnectionConsumer$0(PulsarConnection.java:688)
        at com.datastax.oss.pulsar.jms.PulsarConnection.buildConnectionConsumer(PulsarConnection.java:965)
        at com.datastax.oss.pulsar.jms.PulsarConnection.createConnectionConsumer(PulsarConnection.java:687)
        at org.jboss.resource.adapter.jms.inflow.JmsServerSessionPool.setupConsumer(JmsServerSessionPool.java:289)
        at org.jboss.resource.adapter.jms.inflow.JmsServerSessionPool.start(JmsServerSessionPool.java:89)
        at org.jboss.resource.adapter.jms.inflow.JmsActivation.setupSessionPool(JmsActivation.java:656)
        at org.jboss.resource.adapter.jms.inflow.JmsActivation.setup(JmsActivation.java:359)
        at org.jboss.resource.adapter.jms.inflow.JmsActivation$SetupActivation.run(JmsActivation.java:729)
        at org.jboss.resource.work.WorkWrapper.execute(WorkWrapper.java:205)
        at org.jboss.util.threadpool.BasicTaskWrapper.run(BasicTaskWrapper.java:260)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
        at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:251)
        at org.apache.pulsar.client.admin.internal.TopicsImpl.createSubscription(TopicsImpl.java:1108)
        at com.datastax.oss.pulsar.jms.PulsarConnectionFactory.ensureQueueSubscription(PulsarConnectionFactory.java:932)
        ... 17 more

Here is the complete trace - Its ccalling OMSClientImpl.subscribeAsyncParent at some point - 

javax.jms.JMSException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Receiver queue size needs to be greater than 0 for Topics Consumer

at com.datastax.oss.pulsar.jms.Utils.handleException(Utils.java:63)

at com.datastax.oss.pulsar.jms.PulsarConnectionFactory.ensureQueueSubscription(PulsarConnectionFactory.java:963)

at com.datastax.oss.pulsar.jms.PulsarMessageConsumer.subscribe(PulsarMessageConsumer.java:116)

at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:783)

at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:734)

at com.datastax.oss.pulsar.jms.PulsarSession.createConsumer(PulsarSession.java:73)

at com.datastax.oss.pulsar.jms.PulsarConnection.lambda$createConnectionConsumer$0(PulsarConnection.java:688)

at com.datastax.oss.pulsar.jms.PulsarConnection.buildConnectionConsumer(PulsarConnection.java:965)

at com.datastax.oss.pulsar.jms.PulsarConnection.createConnectionConsumer(PulsarConnection.java:687)

at org.jboss.resource.adapter.jms.inflow.JmsServerSessionPool.setupConsumer(JmsServerSessionPool.java:289)

at org.jboss.resource.adapter.jms.inflow.JmsServerSessionPool.start(JmsServerSessionPool.java:89)

at org.jboss.resource.adapter.jms.inflow.JmsActivation.setupSessionPool(JmsActivation.java:656)

at org.jboss.resource.adapter.jms.inflow.JmsActivation.setup(JmsActivation.java:359)

at org.jboss.resource.adapter.jms.inflow.JmsActivation$SetupActivation.run(JmsActivation.java:729)

at org.jboss.resource.work.WorkWrapper.execute(WorkWrapper.java:205)

at org.jboss.util.threadpool.BasicTaskWrapper.run(BasicTaskWrapper.java:260)

at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

at java.lang.Thread.run(Unknown Source)

Caused by: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Receiver queue size needs to be greater than 0 for Topics Consumer

at org.apache.pulsar.client.api.PulsarClientException.unwrap(PulsarClientException.java:1074)

at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribe(ConsumerBuilderImpl.java:101)

at com.datastax.oss.pulsar.jms.PulsarConnectionFactory.ensureQueueSubscription(PulsarConnectionFactory.java:946)

... 17 more

Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Receiver queue size needs to be greater than 0 for Topics Consumer

at java.util.concurrent.CompletableFuture.reportGet(Unknown Source)

at java.util.concurrent.CompletableFuture.get(Unknown Source)

at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribe(ConsumerBuilderImpl.java:99)

... 18 more

Caused by: java.lang.IllegalArgumentException: Receiver queue size needs to be greater than 0 for Topics Consumer

at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)

at org.apache.pulsar.client.impl.MultiTopicsConsumerImpl.<init>(MultiTopicsConsumerImpl.java:139)

at org.apache.pulsar.client.impl.MultiTopicsConsumerImpl.<init>(MultiTopicsConsumerImpl.java:128)

at org.apache.pulsar.client.impl.MultiTopicsConsumerImpl.<init>(MultiTopicsConsumerImpl.java:112)

at org.apache.pulsar.client.impl.PulsarClientImpl.multiTopicSubscribeAsync(PulsarClientImpl.java:514)

at org.apache.pulsar.client.impl.PulsarClientImpl.subscribeAsync(PulsarClientImpl.java:466)

at com.intuit.platform.messaging.pulsar.client.OMSClientImpl.subscribeAsyncParent(OMSClientImpl.java:328)

at com.intuit.platform.messaging.pulsar.client.OMSClientImpl.subscribeAsyncInternal(OMSClientImpl.java:429)

at com.intuit.platform.messaging.pulsar.client.OMSClientImpl.subscribeAsync(OMSClientImpl.java:533)

at org.apache.pulsar.client.impl.ConsumerBuilderImpl.lambda$subscribeAsync$1(ConsumerBuilderImpl.java:173)

at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source)

at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source)

at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribeAsync(ConsumerBuilderImpl.java:171

--- a/pulsar-jms-cli/README.md
+++ b/pulsar-jms-cli/README.md
@@ -81,5 +81,3 @@ With this command you create a Pulsar subscription on a JMS Destination, and you
 side selector that is applied per subscription.
 
 If you run the `describe` command you will see the new subscription.
-=======
-```

--- a/pulsar-jms-integration-tests/pom.xml
+++ b/pulsar-jms-integration-tests/pom.xml
@@ -97,8 +97,8 @@
             <configuration>
               <tasks>
                 <echo>copy filters</echo>
-                <mkdir dir="${project.build.outputDirectory}/filters" />
-                <copy verbose="true" file="${basedir}/../pulsar-jms-filters/target/pulsar-jms-filters-${project.version}.jar" tofile="${project.build.outputDirectory}/filters/jms-filter.nar" />
+                <mkdir dir="${project.build.outputDirectory}/filters"/>
+                <copy verbose="true" file="${basedir}/../pulsar-jms-filters/target/pulsar-jms-filters-${project.version}.jar" tofile="${project.build.outputDirectory}/filters/jms-filter.nar"/>
               </tasks>
             </configuration>
           </execution>

--- a/pulsar-jms/pom.xml
+++ b/pulsar-jms/pom.xml
@@ -108,8 +108,8 @@
             <configuration>
               <tasks>
                 <echo>copy filters</echo>
-                <mkdir dir="${project.build.outputDirectory}/filters" />
-                <copy verbose="true" file="${basedir}/../pulsar-jms-filters/target/pulsar-jms-filters-${project.version}.jar" tofile="${project.build.outputDirectory}/filters/jms-filter.nar" />
+                <mkdir dir="${project.build.outputDirectory}/filters"/>
+                <copy verbose="true" file="${basedir}/../pulsar-jms-filters/target/pulsar-jms-filters-${project.version}.jar" tofile="${project.build.outputDirectory}/filters/jms-filter.nar"/>
               </tasks>
             </configuration>
           </execution>

--- a/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnection.java
+++ b/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnection.java
@@ -32,9 +32,11 @@ import javax.jms.InvalidClientIDException;
 import javax.jms.InvalidDestinationException;
 import javax.jms.InvalidSelectorException;
 import javax.jms.JMSException;
+import javax.jms.MessageConsumer;
 import javax.jms.Queue;
 import javax.jms.QueueConnection;
 import javax.jms.QueueSession;
+import javax.jms.ServerSession;
 import javax.jms.ServerSessionPool;
 import javax.jms.Session;
 import javax.jms.TemporaryQueue;
@@ -682,7 +684,8 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
       int maxMessages)
       throws JMSException {
     checkNotClosed();
-    throw new JMSException("not supported");
+    return buildConnectionConsumer(
+        sessionPool, (session) -> session.createConsumer(destination, messageSelector));
   }
 
   /**
@@ -727,7 +730,9 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
       int maxMessages)
       throws JMSException {
     checkNotClosed();
-    throw new JMSException("not supported");
+    return buildConnectionConsumer(
+        sessionPool,
+        session -> session.createSharedConsumer(topic, subscriptionName, messageSelector));
   }
 
   /**
@@ -772,7 +777,9 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
       int maxMessages)
       throws JMSException {
     checkNotClosed();
-    throw new JMSException("not supported");
+    return buildConnectionConsumer(
+        sessionPool,
+        session -> session.createDurableConsumer(topic, subscriptionName, messageSelector, false));
   }
 
   /**
@@ -817,7 +824,9 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
       int maxMessages)
       throws JMSException {
     checkNotClosed();
-    throw new JMSException("not supported");
+    return buildConnectionConsumer(
+        sessionPool,
+        session -> session.createSharedDurableConsumer(topic, subscriptionName, messageSelector));
   }
 
   public void unregisterSession(PulsarSession session) {
@@ -932,8 +941,10 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
 
   @Override
   public ConnectionConsumer createConnectionConsumer(
-      Topic topic, String s, ServerSessionPool serverSessionPool, int i) throws JMSException {
-    throw new JMSException("Not implemented");
+      Topic topic, String messageSelector, ServerSessionPool serverSessionPool, int i)
+      throws JMSException {
+    return buildConnectionConsumer(
+        serverSessionPool, (session) -> session.createConsumer(topic, messageSelector));
   }
 
   void removeTemporaryDestination(PulsarTemporaryDestination pulsarTemporaryDestination) {
@@ -942,5 +953,23 @@ public class PulsarConnection implements Connection, QueueConnection, TopicConne
 
   public String getConnectionId() {
     return connectionId;
+  }
+
+  @FunctionalInterface
+  private interface ConsumerBuilder {
+    MessageConsumer build(Session session) throws JMSException;
+  }
+
+  private ConnectionConsumer buildConnectionConsumer(
+      ServerSessionPool sessionPool, ConsumerBuilder consumerBuilder) throws JMSException {
+    ServerSession serverSession = sessionPool.getServerSession();
+    Session session = serverSession.getSession();
+    // this session is supposed to have a MessageListener that receives the messages
+    MessageConsumer consumer = consumerBuilder.build(session);
+    consumer.setMessageListener(session.getMessageListener());
+    ConnectionConsumer connectionConsumer =
+        new PulsarConnectionConsumer((PulsarMessageConsumer) consumer, sessionPool);
+    serverSession.start();
+    return connectionConsumer;
   }
 }

--- a/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnectionConsumer.java
+++ b/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnectionConsumer.java
@@ -1,0 +1,46 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.jms;
+
+import javax.jms.ConnectionConsumer;
+import javax.jms.IllegalStateException;
+import javax.jms.JMSException;
+import javax.jms.ServerSessionPool;
+
+class PulsarConnectionConsumer implements ConnectionConsumer {
+  private final PulsarMessageConsumer consumer;
+  private final ServerSessionPool serverSessionPool;
+  private volatile boolean closed;
+
+  PulsarConnectionConsumer(PulsarMessageConsumer consumer, ServerSessionPool serverSessionPool) {
+    this.consumer = consumer;
+    this.serverSessionPool = serverSessionPool;
+  }
+
+  @Override
+  public ServerSessionPool getServerSessionPool() throws JMSException {
+    if (closed) {
+      throw new IllegalStateException("The Connection Consumer is closed");
+    }
+    return serverSessionPool;
+  }
+
+  @Override
+  public void close() throws JMSException {
+    closed = true;
+    consumer.close();
+  }
+}

--- a/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnectionFactory.java
+++ b/pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/PulsarConnectionFactory.java
@@ -87,6 +87,7 @@ public class PulsarConnectionFactory
   private SubscriptionType topicSharedSubscriptionType = SubscriptionType.Shared;
   private long waitForServerStartupTimeout = 60000;
   private boolean usePulsarAdmin = true;
+  private int precreateQueueSubscriptionConsumerQueueSize = 0;
   private boolean initialized;
   private boolean closed;
 
@@ -233,6 +234,11 @@ public class PulsarConnectionFactory
 
       this.usePulsarAdmin =
           Boolean.parseBoolean(getAndRemoveString("jms.usePulsarAdmin", "true", configuration));
+
+      this.precreateQueueSubscriptionConsumerQueueSize =
+          Integer.parseInt(
+              getAndRemoveString(
+                  "jms.precreateQueueSubscriptionConsumerQueueSize", "0", configuration));
 
       final String rawTopicSharedSubscriptionType =
           getAndRemoveString(
@@ -869,7 +875,7 @@ public class PulsarConnectionFactory
               .subscriptionType(getTopicSharedSubscriptionType())
               .subscriptionName(getQueueSubscriptionName(destination))
               .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
-              .receiverQueueSize(0)
+              .receiverQueueSize(getPrecreateQueueSubscriptionConsumerQueueSize())
               .topic(fullQualifiedTopicName)
               .subscribe()
               .close();
@@ -1196,5 +1202,9 @@ public class PulsarConnectionFactory
 
   public synchronized boolean isClosed() {
     return closed;
+  }
+
+  private synchronized int getPrecreateQueueSubscriptionConsumerQueueSize() {
+    return precreateQueueSubscriptionConsumerQueueSize;
   }
 }

--- a/pulsar-jms/src/test/java/com/datastax/oss/pulsar/jms/ConnectionConsumerTest.java
+++ b/pulsar-jms/src/test/java/com/datastax/oss/pulsar/jms/ConnectionConsumerTest.java
@@ -1,0 +1,190 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.jms;
+
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.datastax.oss.pulsar.jms.utils.PulsarCluster;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.jms.Connection;
+import javax.jms.ConnectionConsumer;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.MessageProducer;
+import javax.jms.ServerSession;
+import javax.jms.ServerSessionPool;
+import javax.jms.Session;
+import javax.jms.Topic;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+public class ConnectionConsumerTest {
+
+  @TempDir public static Path tempDir;
+  private static PulsarCluster cluster;
+
+  @BeforeAll
+  public static void before() throws Exception {
+    cluster = new PulsarCluster(tempDir);
+    cluster.start();
+  }
+
+  @AfterAll
+  public static void after() throws Exception {
+    if (cluster != null) {
+      cluster.close();
+    }
+  }
+
+  private interface ConnectionConsumerBuilder {
+    ConnectionConsumer build(
+        Connection connection,
+        Destination destination,
+        String selector,
+        ServerSessionPool serverSessionPool)
+        throws Exception;
+  }
+
+  @Test
+  public void createConnectionConsumerQueue() throws Exception {
+    simpleTest(
+        (Connection connection,
+            Destination destination,
+            String selector,
+            ServerSessionPool serverSessionPool) -> {
+          return connection.createConnectionConsumer(destination, selector, serverSessionPool, 10);
+        },
+        false);
+  }
+
+  @Test
+  public void createConnectionConsumerTopic() throws Exception {
+    simpleTest(
+        (Connection connection,
+            Destination destination,
+            String selector,
+            ServerSessionPool serverSessionPool) -> {
+          return connection.createConnectionConsumer(destination, selector, serverSessionPool, 10);
+        },
+        true);
+  }
+
+  @Test
+  public void createDurableConnectionConsumer() throws Exception {
+    simpleTest(
+        (Connection connection,
+            Destination destination,
+            String selector,
+            ServerSessionPool serverSessionPool) -> {
+          return connection.createDurableConnectionConsumer(
+              (Topic) destination, "subname", selector, serverSessionPool, 10);
+        },
+        true);
+  }
+
+  @Test
+  public void createSharedConnectionConsumer() throws Exception {
+    simpleTest(
+        (Connection connection,
+            Destination destination,
+            String selector,
+            ServerSessionPool serverSessionPool) -> {
+          return connection.createSharedConnectionConsumer(
+              (Topic) destination, "subname", selector, serverSessionPool, 10);
+        },
+        true);
+  }
+
+  @Test
+  public void createSharedDurableConnectionConsumer() throws Exception {
+    simpleTest(
+        (Connection connection,
+            Destination destination,
+            String selector,
+            ServerSessionPool serverSessionPool) -> {
+          return connection.createSharedDurableConnectionConsumer(
+              (Topic) destination, "subname", selector, serverSessionPool, 10);
+        },
+        true);
+  }
+
+  private void simpleTest(ConnectionConsumerBuilder builder, boolean topic) throws Exception {
+    Map<String, Object> properties = new HashMap<>();
+    properties.put("webServiceUrl", cluster.getAddress());
+    properties.put("jms.enableClientSideEmulation", true);
+    // required for createDurableConnectionConsumer
+    properties.put("jms.clientId", "test");
+
+    try (PulsarConnectionFactory factory = new PulsarConnectionFactory(properties);
+        Connection connection = factory.createConnection();
+        Session session = connection.createSession(Session.AUTO_ACKNOWLEDGE); ) {
+      connection.start();
+      Destination destination =
+          topic
+              ? session.createTopic("persistent://public/default/test-" + UUID.randomUUID())
+              : session.createQueue("persistent://public/default/test-" + UUID.randomUUID());
+      SimpleMessageListener listener = new SimpleMessageListener();
+      session.setMessageListener(listener);
+
+      AtomicBoolean startCalled = new AtomicBoolean();
+      ServerSessionPool serverSessionPool =
+          new ServerSessionPool() {
+            @Override
+            public ServerSession getServerSession() throws JMSException {
+              return new ServerSession() {
+                @Override
+                public Session getSession() throws JMSException {
+                  return session;
+                }
+
+                @Override
+                public void start() throws JMSException {
+                  startCalled.set(true);
+                }
+              };
+            }
+          };
+
+      ConnectionConsumer connectionConsumer =
+          builder.build(connection, destination, "TRUE", serverSessionPool);
+
+      assertTrue(startCalled.get());
+
+      try (MessageProducer producer = session.createProducer(destination); ) {
+        for (int i = 0; i < 10; i++) {
+          producer.send(session.createTextMessage("foo-" + i));
+        }
+      }
+      // wait for messages to arrive
+      await().until(listener.receivedMessages::size, equalTo(10));
+
+      for (int i = 0; i < 10; i++) {
+        assertEquals("foo-" + i, listener.receivedMessages.get(i).getBody(String.class));
+      }
+
+      connectionConsumer.close();
+    }
+  }
+}

